Use Webdav PUT for uploads

PVince81 opened the pull request:
Fixes https://github.com/owncloud/core/issues/4380
- [x] send mtime with upload (bonus feature!)
- [x] conflict handling
  - [x] replace file after conflict resolution
  - [x] autorename
  - [x] abort
- [x] error cases
  - [x] parent folder does not exist any more
- [x] folder upload
- [x] highlight all files after upload (code currently discarded)
- ~~make ajax/upload.php as close as possible to webdav upload (for IE8 to reduce code paths)~~
- [x] Webdav plugin for iframe transport
- [x] delete ajax/upload.php
- [x] clean up messy code (a bit)
- [x] unit tests
  - [x] file list
  - [x] PHP side
  - [x] file-upload.js
#### Test cases
- [x] TEST: create file from menu + error cases
- [x] TEST: create folder from menu + error cases
- [x] TEST: upload files into current folder
- [x] TEST: upload files into a subfolder
- [x] TEST: upload files to breadcrumb
- [x] TEST: upload folder with multiple subdirs (Chrome only)
- [x] TEST: upload file/folder into a subfolder that does not exist (renamed/deleted concurrently)
- [x] TEST: cancel upload
- [x] TEST: conflict dialog
  - [x] TEST: autorename
  - [x] TEST: overwrite
  - [x] TEST: don't overwrite
- [x] TEST: in all modern browsers
- [x] TEST: IE8 :scream: 
- [x] TEST: all upload scenarios in public page (with and without password)
- [x] TEST: special characters

Commit:
Use Webdav PUT for uploads in the web browser

- uses PUT method with jquery.fileupload for regular and public file
  lists
- for IE and browsers that don't support it, use POST with iframe
  transport
- implemented Sabre plugin to handle iframe transport and redirect the
  embedded PUT request to the proper handler
- added RFC5995 POST to file collection with "add-member" property to
  make it possible to auto-rename conflicting file names
- remove obsolete ajax/upload.php and obsolete ajax routes

MorrisJobke commented on apps/files/js/file-upload.js:
```diff
@@ -518,7 +524,7 @@ OC.Upload = {
 			};
 
-			// initialize jquery fileupload (blueimp)
+			// initialize jquery fileupload (nlueimp)
```
really? :wink:

PVince81 replied:
whoops

PVince81 commented on apps/files/js/file-upload.js:
```diff
@@ -68,17 +68,260 @@ function addFormData(formData, newData) {
+			}
+
+			parentPromise.then(function() {
+				console.log('createDirectory: ' + fullPath);
```
eeergh

PVince81 commented on apps/dav/lib/connector/sabre/iframetransportplugin.php:
```diff
@@ -0,0 +1,201 @@
+ * mostly used as a workaround for browsers that
+ * do not support PUT upload.
+ */
+class IFrameTransportPlugin extends \Sabre\DAV\ServerPlugin {
```
@DeepDiver1975 @evert is this acceptable or have a gone too far :wink: 

I wanted to replace the old ajax/upload.php and use Webdav PUT for uploads.
However old browsers like IE8/IE9 need an iframe transport and do a POST, so this plugin catches that and forwards the call to `createFile` / `updateFile` and checks preconditions.

Goal is to bring the code paths as close as possible.

What do you think ?

PVince81 replied:
also @LukasReschke in case of security considerations. I added the CSRF check here, but we can probably move it to a generic plugin or the auth plugin.

evert replied:
Don't really see anything wrong with it. You could even take it one step further by creating a $subRequest / $subResponse and sending that to Server::invokeMethod as a more generic approach.

PVince81 replied:
Ah! That's the kind of thing I was looking for :smile: Thanks!
Problem though is that I need to post-process the response to repackage it as JSON.
I'll have a look if that's possible.

PVince81 commented on apps/dav/lib/connector/sabre/iframetransportplugin.php:
```diff
@@ -0,0 +1,201 @@
+		if (!is_string($data)) {
+			$data = json_encode($data);
+		}
+		$response->setBody(json_encode([
```
since we're in an iframe, we need to append the response in the body. The JS code will pluck it out.

PVince81 commented on apps/dav/lib/connector/sabre/iframetransportplugin.php:
```diff
@@ -0,0 +1,188 @@
+		$request->setBody($resource);
+		$request->setMethod($method);
+
+		$this->server->invokeMethod($request, $response, false);
```
@evert here we go, the more generic `invokeMethod` approach :smile: 

So far we're only interested it the PUT method, but this could be reused in the future if at some point we'd like to have cross-domain requests using iframe transport. :smile:

PVince81 replied:
I think this is also the right moment to say that I love Sabre's plugin approach for everything!

evert replied:
Awesome =) great to hear that

PVince81 commented:
Will do autorename using Webdav POST on the collection with suggested name, as per RFC 5995 http://greenbytes.de/tech/webdav/rfc5995.html (starting to like that one!)

PVince81 commented:
And here we go, autorename implemented based on RFC 5995. So cool!

- [ ] BUG: conflict dialog must appear at the end of all transfers

Commit:
JS utility function to compare paths

PVince81 commented:
Rebased.

Now that the uploader class is a bit more encapsulated, I'd like to add a bit more unit tests:
- [ ] more JS unit tests for uploader
- [ ] PHP unit tests for the new sabre functions
   - [ ] iframe transport plugin
   - [ ] POST+slug to collection

PVince81 commented:
I'm still wondering whether it's ok to have all these extra PROPFIND after upload, or whether we should try and return all the needed info through headers...

PVince81 commented:
Here are some ideas to remove the extra PROPFIND on after upload:
- get fileid from headers
- get etag from headers
- mtime? if accepted => use sent one, else set to now
- size? use one if browser knows it, else need to return as header
- share/tags info => not really needed or can be computed
- mimetype ? could be problematic, unless we return it too...
- permissions: compute ?
- mount type: compute ?

PVince81 commented:
Would be even easier if we could just return a JSON blob (or the result of XML PROPFIND) as a result of the PUT request.
@evert would that violate any Webdav convention ?

evert commented:
I don't think there's any problems with doing stuff with the response to a `PUT` request. It's a bit unusual

PVince81 commented:
@evert so I guess it might be better to return information as headers instead ? That's what we party did so far: `OC-FileId` contains the file id and `OC-Mtime: accepted` tells whether the server accepted the mtime. We could add `OC-Permissions`, `OC-Size`.

Keeping the subsequent PROPFIND is so easy... but some people might complain about performance.

evert commented:
Well, there's another idea.. you could add a specific header to `PUT` requests indicating that you want the server to return information about the resource... But yea headers would work as well.

PVince81 commented:
- [ ] BUG: IE8 requests basic auth when uploading to public page

PVince81 commented:
IE8 will not be supported for 9.0, see https://github.com/owncloud/core/issues/15567#issuecomment-166754301

- [ ] investigate whether iframe transport is needed for IE9+ and others like Safari

evert commented:
:tada:

PVince81 commented:
Noooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo...

IE9 also requires the iframe transport.

PVince81 commented:
Did not have time to finish this and I'm worried about the extra PROPFIND calls after each upload.
Ideal would be to somehow make the PUT request return some metadata and/or reuse whichever metadata is already known in the web UI.

Moving to 9.1

oparoz commented:
Cancelling an upload on folder does not restore the folder to its original state:
1. Upload some files in the current folder
1. Create a new folder
1. Upload the same files in the new folder
1. Hit cancel in the conflict dialogue (yes, it shows up because it only compares file names)

PVince81 commented:
I will finish this for 9.2, hopefully...

PVince81 commented:
- [ ] TEST: retest with "isReceivedShare" logic when doing a PUT on a received share file while current user has zero quota

PVince81 commented:
Rebased. Unfortunately the new jquery version had an infinite loop, so had to change some pieces to avoid it.

In general I'm still not happy with the current code in file-uploads.js, still bloated...

PVince81 commented:
Considering that starting with 9.1 we only support IE11+, let's see if we can finally drop iframe transport support...

Commit:
Goodbye Iframe transport !

Not needed any more in IE >= 11

PVince81 commented:
Looks like implementing chunking/resume will be a PITA. 
The current lib uses the "Content-Range" header with PUT, which is forbidden https://github.com/blueimp/jQuery-File-Upload/wiki/Chunked-file-uploads#how-do-chunked-uploads-work 

Would need to find a way to hack the lib to be able to implement chunk sending in our own way using the new chunked upload. The alternative, resumable.js, would require replacing the current lib completely.

Commit:
Add support for chunked upload

Hacked around Blueimp's jquery.fileupload to make it work with our new
chunking API.

PVince81 commented:
Added support for chunked upload using the new chunking endpoint.

Chunk size: 10 MB.

Known issues:
- [ ] BUG: progress bar disappears too early at the end of a chunk upload (will need more hackery to delay the "end of transfer" event)

While testing I found this perf issue: https://github.com/owncloud/core/issues/25493

PVince81 commented:
- [ ] TEST: test how blueimp deals with failed chunks and whether resuming works properly

DebianRoxx commented:
Why not an "intelligent" mode for the desktop client ?
If the file doesn't exists on owncloud, use same upload method than web client else use chunk method.

PVince81 commented:
Okay, enough fooling around with chunking.

I'll modify this PR to be only about Webdav PUT for now, with subsequent PROPFINDs to retrieve the info of the files that were just uploaded.

Then we can look into getting rid of these extra PROPFINDs  and also how to do parallel chunking in a separate PR (which might require even switching the upload library).

PVince81 commented:
Rebased. Some JS tests will fail, need to look into it.

PVince81 commented:
The plan is as follows:
- keep the chunking commit as it refactors stuff a bit further and makes it better, but chunking is disabled (one line can enable it)
- [ ] fix those JS tests
- [ ] retest all upload scenarios listed in the original post
- [ ] motivate people to review this PR

Commit:
Fix js unit tests for webdav put upload changes

Commit:
More fixes to file upload

PVince81 commented:
- [x] Tested in Firefox, Chromium and IE11.

Ready for review @DeepDiver1975 @butonic @owncloud/qa

butonic commented:
Really nice! 👍 

That being said I cannot even login in IE8:
```
Webpage error details

User Agent: Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; Trident/4.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C)
Timestamp: Mon, 5 Sep 2016 08:42:47 UTC


Message: Object doesn't support this property or method
Line: 2
Char: 29759
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/vendor/jquery/dist/jquery.min.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'jQuery' is undefined
Line: 2
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/vendor/jquery-migrate/jquery-migrate.min.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'jQuery' is undefined
Line: 6
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/vendor/jquery-ui/ui/jquery-ui.custom.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'jQuery' is undefined
Line: 11
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/vendor/bootstrap/js/tooltip.js?v=14269cbd8477654a15c200a1df0a28b2


Message: Expected identifier, string or number
Line: 372
Char: 13
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/vendor/clipboard/dist/clipboard.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'jQuery' is undefined
Line: 1
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/jquery.ocdialog.js?v=14269cbd8477654a15c200a1df0a28b2


Message: Object expected
Line: 1638
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/js.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'jQuery' is undefined
Line: 51
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/octemplate.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'OCA' is undefined
Line: 389
Char: 2
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/search/js/search.js?v=14269cbd8477654a15c200a1df0a28b2


Message: Object expected
Line: 1
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/oc-requesttoken.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'jQuery' is undefined
Line: 11
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/apps.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'jQuery' is undefined
Line: 62
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/placeholder.js?v=14269cbd8477654a15c200a1df0a28b2


Message: 'jQuery' is undefined
Line: 49
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/jquery.avatar.js?v=14269cbd8477654a15c200a1df0a28b2


Message: Object expected
Line: 23
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/backgroundjobs.js?v=14269cbd8477654a15c200a1df0a28b2


Message: Object expected
Line: 444
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/share.js?v=14269cbd8477654a15c200a1df0a28b2


Message: Object expected
Line: 2
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/visitortimezone.js?v=14269cbd8477654a15c200a1df0a28b2


Message: Object expected
Line: 143
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/lostpassword.js?v=14269cbd8477654a15c200a1df0a28b2


Message: Object expected
Line: 31
Char: 1
Code: 0
URI: https://192.168.56.56/Repositories/oc/core/core/js/login.js?v=14269cbd8477654a15c200a1df0a28b2

```
I reverted to 83083488815e to see if it was caused to to move to make, but I get the same errors there. I tried IE11 in IE8 emulation and the Win7 IE8 VM: https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/

@DeepDiver1975 @PVince81 @felixboehm de we still support IE8?

DeepDiver1975 commented:
Ie8 support is long gone

DeepDiver1975 commented:
That said - we should prevent users from logging in with unsupported browsers

butonic commented:
Ok, IE9, uploading a 300mb video fails:
```
SCRIPT5007: Unable to get value of the property '0': object is null or undefined 
file-upload.js?v=14269cbd8477654a15c200a1df0a28b2, line 287 character 4
```

davitol commented:
@butonic In oC 9.1 IE9 is no longer supported. Here are the supported web browsers for oC 9.1. 

Web browser: IE11+ (except Compatibility Mode), Firefox 14+, Chrome 18+, Safari 5+

https://doc.owncloud.org/server/9.1/admin_manual/installation/system_requirements.html

butonic commented:
I scanned the thread from the start and stumbled over IE8 and IE9 ... reading thoroughly would have helped: https://github.com/owncloud/core/pull/21237#issuecomment-232924894

sorry for the noise

so a 👍 from me

PVince81 commented:
@butonic yeah, because when I started we still wanted to support IE8 and I had a workaround with an iframe transport plugin... then killed it again when we killed IE9.

Thanks for the reviews.
